File: services/graphql-faas/src/graphql/types/Transaction.js
const {
  GraphQLObjectType,
  GraphQLInputObjectType,
  GraphQLNonNull,
  GraphQLString,
  GraphQLID
} = require('graphql')

const TransactionType = new GraphQLObjectType({
  name: 'Transaction',
  description: 'fields on transactions',
  fields: {
    id: {
      type: GraphQLID,
      description: 'transaction item id'
    },
    debitor: {
      type: GraphQLString,
      description: 'debitor account sending payment'
    },
    debitor_profile_latlng: {
      type: GraphQLString,
      description: 'lat long of debitor account profile'
    },
    debitor_transaction_latlng: {
      type: GraphQLString,
      description: 'lat long of debitor transaction'
    },
    debitor_approval_time: {
      type: GraphQLString,
      description: 'time debitor approved transaction'
    },
    debitor_device: {
      type: GraphQLString,
      description: 'device used by debitor for transaction'
    },
    creditor: {
      type: GraphQLString,
      description: 'creditor account; receiving payment'
    },
    creditor_profile_latlng: {
      type: GraphQLString,
      description: 'lat long of creditor account profile'
    },
    creditor_transaction_latlng: {
      type: GraphQLString,
      description: 'lat long of creditor transaction'
    },
    creditor_approval_time: {
      type: GraphQLString,
      description: 'time creditor approved transaction'
    },
    creditor_device: {
      type: GraphQLString,
      description: 'device used by creditor for transaction'
    },
    name: {
      type: GraphQLString,
      description: 'name of good or service'
    },
    price: {
      type: GraphQLString,
      description: 'value of good or service'
    },
    quantity: {
      type: GraphQLString,
      description: 'quantity of good or service'
    },
    unit_of_measurement: {
      type: GraphQLString,
      description: 'unit of measurement of good or service'
    },
    units_measured: {
      type: GraphQLString,
      description: 'units measured of good or service'
    },
    rule_instance_id: {
      type: GraphQLString,
      description: 'id of rule instance automatically adding transaction item'
    },
    transaction_id: {
      type: GraphQLString,
      description: 'id shared by goods and services grouped in a transaction'
    },
    author: {
      type: GraphQLString,
      description: 'initial author of transaction'
    },
    expiration_time: {
      type: GraphQLString,
      description: 'time incomplete transaction expires'
    },
    creditor_rejection_time: {
      type: GraphQLString,
      description: 'time transaction rejected by creditor'
    },
    debitor_rejection_time: {
      type: GraphQLString,
      description: 'time transaction rejected by debitor'
    }
  }
})

const TransactionInputType = new GraphQLInputObjectType({
  name: 'TransactionInputType',
  type: TransactionType,
  fields: {
    id: { type: GraphQLID },
    debitor: { type: new GraphQLNonNull(GraphQLString) },
    debitor_profile_latlng: { type: GraphQLString },
    debitor_transaction_latlng: { type: GraphQLString },
    debitor_device: { type: GraphQLString },
    creditor: { type: new GraphQLNonNull(GraphQLString) },
    creditor_profile_latlng: { type: GraphQLString },
    creditor_transaction_latlng: { type: GraphQLString },
    creditor_device: { type: GraphQLString },
    name: { type: new GraphQLNonNull(GraphQLString) },
    price: { type: new GraphQLNonNull(GraphQLString) },
    quantity: { type: new GraphQLNonNull(GraphQLString) },
    unit_of_measurement: { type: GraphQLString },
    units_measured: { type: GraphQLString },
    rule_instance_id: { type: GraphQLString },
    transaction_id: { type: GraphQLString },
    author: { type: new GraphQLNonNull(GraphQLString) },
    expiration_time: { type: GraphQLString },
    creditor_rejection_time: { type: GraphQLString },
    debitor_rejection_time: { type: GraphQLString }
  }
})

module.exports = {
  TransactionType,
  TransactionInputType,
}

File: services/req-query-account-faas/transaction.go
package faas

import (
	"encoding/json"
	"log"

	"gopkg.in/guregu/null.v4"
)

// Event defines a lambda event
type Event struct {
	TransactionID        string `json:"transaction_id"`
	Account              string `json:"account"`
	GraphQLRequestSender string `json:"graphqlRequestSender"`
}

// Transaction defines a transaction
type Transaction struct {
	TableID                   int32       `json:"id,string"`
	Name                      string      `json:"name"`
	Quantity                  string      `json:"quantity"`
	Price                     string      `json:"price"`
	UnitOfMeasurement         null.String `json:"unit_of_measurement"`
	UnitsMeasured             null.String `json:"units_measured"`
	RuleInstanceID            null.String `json:"rule_instance_id"`
	TransactionID             null.String `json:"transaction_id"`
	Author                    null.String `json:"author"`
	ExpirationTime            null.String `json:"expiration_time"`
	Debitor                   null.String `json:"debitor"`
	Creditor                  null.String `json:"creditor"`
	DebitorProfileLatLng      null.String `json:"debitor_profile_latlng"`
	CreditorProfileLatlng     null.String `json:"creditor_profile_latlng"`
	DebitorTransactionLatLng  null.String `json:"debitor_transaction_latlng"`
	CreditorTransactionLatLng null.String `json:"creditor_transaction_latlng"`
	DebitorApprovalTime       null.String `json:"debitor_approval_time"`
	CreditorApprovalTime      null.String `json:"creditor_approval_time"`
	DebitorDevice             null.String `json:"debitor_device"`
	CreditorDevice            null.String `json:"creditor_device"`
	DebitApprover             null.String `json:"debit_approver"`
	CreditApprover            null.String `json:"credit_approver"`
	CreditorRejectionTime     null.String `json:"creditor_rejection_time"`
	DebitorRejectionTime      null.String `json:"debitor_rejection_time"`
	CreatedAt                 null.String `json:"createdAt"`
}

// Transactions ...
type Transactions []*Transaction

// Marshal returns transactions as json
func (ts Transactions) Marshal() (string, error) {
	j, err := json.Marshal(ts)
	if err != nil {
		return "", err
	}

	js := string(j)

	// ### debugging only
	transInd, err := json.MarshalIndent(ts, "", "  ")
	if err != nil {
		log.Print(err)
		return "", err
	}
	indented := string(transInd)
	_ = indented
	// log.Print(indented)
	// ###

	return js, nil
}
